gensim/models/fasttext_inner.pyx

```diff
@@ -0,0 +1,544 @@
+        for token in sent:
+            word = vlookup[token] if token in vlookup else None
+            if word is None:
+                continue  # leaving `effective_words` unchanged = shortening the sentence = expanding the window
+            if sample and word.sample_int < random_int32(&next_random):
+                continue
+            indexes[effective_words] = word.index
+
+            subwords = [model.wv.ngrams[subword_i] for subword_i in model.wv.ngrams_word[model.wv.index2word[word.index]]]
+            word_subwords = np.array([word.index] + subwords, dtype=np.uint32)
+            subwords_idx_len[effective_words] = <int>(len(subwords) + 1)
+            subwords_idx[effective_words] = <np.uint32_t *>np.PyArray_DATA(word_subwords)
+            # ensures reference count of word_subwords doesn't reach 0
+            subword_arrays[effective_words] = word_subwords
+
+            if hs:
+                codelens[effective_words] = <int>len(word.code)
+                codes[effective_words] = <np.uint8_t *>np.PyArray_DATA(word.code)
+                points[effective_words] = <np.uint32_t *>np.PyArray_DATA(word.point)
+
+            effective_words += 1
+            if effective_words == MAX_SENTENCE_LEN:
+                break  # TODO: log warning, tally overflow?
+
+        # keep track of which words go into which sentence, so we don't train
+        # across sentence boundaries.
+        # indices of sentence number X are between <sentence_idx[X], sentence_idx[X])
+        effective_sentences += 1
+        sentence_idx[effective_sentences] = effective_words
+
+        if effective_words == MAX_SENTENCE_LEN:
+            break  # TODO: log warning, tally overflow?
+
+    # precompute "reduced window" offsets in a single randint() call
+    for i, item in enumerate(model.random.randint(0, window, effective_words)):
+        reduced_windows[i] = item
+
+    with nogil:
+        for sent_idx in range(effective_sentences):
+            idx_start = sentence_idx[sent_idx]
+            idx_end = sentence_idx[sent_idx + 1]
+            for i in range(idx_start, idx_end):
+                j = i - window + reduced_windows[i]
+                if j < idx_start:
+                    j = idx_start
+                k = i + window + 1 - reduced_windows[i]
+                if k > idx_end:
+                    k = idx_end
+                for j in range(j, k):
+                    if j == i:
+                        continue
+                    if hs:
+                        fast_sentence_sg_hs(points[j], codes[j], codelens[j], syn0_vocab, syn0_ngrams, syn1, size, subwords_idx[i], subwords_idx_len[i], _alpha, work, l1, word_locks_vocab, word_locks_ngrams)
+                    if negative:
+                        next_random = fast_sentence_sg_neg(negative, cum_table, cum_table_len, syn0_vocab, syn0_ngrams, syn1neg, size, indexes[j], subwords_idx[i], subwords_idx_len[i], _alpha, work, l1, next_random, word_locks_vocab, word_locks_ngrams)
```

Comment:
I don't have a hard limit for line length, but this is way too long. IMO, it's a good idea to keep line length within 100 chars. Going a little over maybe OK where difficult to break, but this is clearly not the case.